feat: add support for AF_UNIX socket path via sockaddr_un

AshishNaware opened the pull request:
<!--
Thanks for contributing! Please ensure your pull request adheres to the following guidelines:

- [ ] All commits contain a well written commit message and are signed-off (see [Submitting a pull request](https://tetragon.io/docs/contribution-guide/submitting-a-pull-request/)).
- [ ] All code is covered by unit and/or end-to-end tests where feasible.
- [ ] All generated files are updated if needed (see [Making changes](https://tetragon.io/docs/contribution-guide/making-changes/)).
- [ ] Provide a title or release-note blurb suitable for the release notes (see [guidelines](https://tetragon.io/docs/contribution-guide/release-notes/#release-note-blurb-in-pr)).
- [ ] Update documentation and write an upgrade note if needed (see [guidelines](https://tetragon.io/docs/contribution-guide/release-notes/#upgrade-notes)).
- [ ] Are you a user of Tetragon? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md) in the Cilium repository.
-->

Fixes <!-- #issue-number -->
#3836 

### Description
<!-- Please describe quickly the change but most importantly the reason or context of your change -->

This PR adds the support for  [sockaddr_un](https://elixir.bootlin.com/linux/v5.15.185/source/include/uapi/linux/un.h#L9) alongside the existing sockaddr_in logic. While Tetragon previously added support for sockaddr_in (#3358), this enhancement extends visibility to UNIX domain sockets, enabling detection and filtering on socket paths such as:

```
/run/containerd/containerd.sock
/var/run/nscd/socket
/dev/log
```
Selectors on `sockaddr_un` arguments support `Equal`, `NotEqual`, `Prefix`, `NotPrefix` on the path, and `Family` 
Note: This feature is available on linux 5.11 or newer.

### Changelog
<!-- Enter the release note text in the codeblock below if needed or remove this section! -->

```release-note
- Added decoding support for `sockaddr_un` in `security_socket_connect` events.
  - Tetragon can now extract and emit UNIX socket paths from AF_UNIX addresses.
  - Supports observability into socket-based interactions like containerd.sock and docker.sock.
  - Selectors support Family and Equal, NotEqual, Prefix, NotPrefix on the path
  - Requires linux 5.11+
```

AshishNaware commented:
This PR is still work in progress as I'm currently addressing the test failures, but I’d appreciate early feedback on the overall struct parsing and decoding approach.

kevsecurity commented on bpf/process/types/sockaddr.h:
```diff
@@ -13,34 +13,52 @@ struct sockaddr_in_type {
+		probe_read(&event->in.sin_port, sizeof(event->in.sin_port), _(&ipv6->sin6_port));
+		break;
+	case AF_UNIX:
+		probe_read_str(event->un.sun_path, sizeof(event->un.sun_path), un->sun_path);
```
This doesn't feel safe as sockaddrs are often user-controlled, and the sun_path might be 108 chars long and not null-terminated (if I've understood correctly).

Also, maybe we should support abstract sockets where the first byte is 0x00, and path name can contain null bytes?

kevsecurity commented on pkg/sensors/tracing/args_linux.go:
```diff
@@ -263,18 +264,43 @@ func getArg(r *bytes.Reader, a argPrinter) api.MsgGenericKprobeArg {
 		arg.Label = a.label
 		return arg
 	case gt.GenericSockaddrType:
-		var address api.MsgGenericKprobeSockaddr
+		var family uint16
+		if err := binary.Read(r, binary.LittleEndian, &family); err != nil {
+			logger.GetLogger().Warn("failed to read sockaddr family", logfields.Error, err)
+			return nil
+		}
+
+		if _, err := r.Seek(6, io.SeekCurrent); err != nil {
+			logger.GetLogger().Warn("failed to skip sockaddr padding", logfields.Error, err)
+			return nil
+		}
+
 		var arg api.MsgGenericKprobeArgSockaddr
+		arg.Family = family
+
+		switch family {
+		case unixsys.AF_INET, unixsys.AF_INET6:
+			var in api.SockaddrIn
+			if err := binary.Read(r, binary.LittleEndian, &in); err != nil {
+				logger.GetLogger().Warn("failed to read sockaddr_in", logfields.Error, err)
+				return nil
+			}
+			arg.SinAddr = network.GetIP(in.SinAddr, family).String()
+			arg.SinPort = uint32(in.SinPort)
+
+		case unixsys.AF_UNIX:
+			var un api.SockaddrUn
+			if err := binary.Read(r, binary.LittleEndian, &un); err != nil {
+				logger.GetLogger().Warn("failed to read sockaddr_un", logfields.Error, err)
+				return nil
+			}
+			arg.Path = string(bytes.Trim(un.SunPath[:], "\x00"))
```
This feels a bit brittle. I feel that we could add a new type (gt.GenericSockaddrUnType) and rename the existing gt.GenericSockaddrType to gt.GenericSockaddrInType, then we can just binary.Read() directly into an appropriate matching data structures in the appropriate cases.

kevsecurity commented on api/v1/tetragon/tetragon.proto:
```diff
@@ -355,11 +355,24 @@ message KprobeSkb {
   uint32 port = 3;
   string path = 4;
+
+  SockaddrIn in = 5;
+  SockaddrUn un = 6;
```
Why are we adding these?

kevsecurity reviewed (changes requested):
Thanks for adding this. Overall I think it makes sense, but I'm concerned about how we interpret the buffer in the BPF code (let's be more robust here), and I think we should probably add a new message type (from BPF to user space) instead of overloading the one supporting sockaddr_in/6.

kevsecurity commented:
Also, as well as passing existing tests, please could you add some tests for the new functionality?

olsajiri reviewed (changes requested):
please separate the change into several commits
typically at least the generated changes go in separate commit, but also single logic change should go in separated commit if possible,  thanks

kevsecurity commented on pkg/sensors/tracing/args_linux.go:
```diff
@@ -276,6 +276,19 @@ func getArg(r *bytes.Reader, a argPrinter) api.MsgGenericKprobeArg {
+
+		arg.Index = uint64(a.index)
+		arg.Family = sockaddr.Family
+		arg.Path = string(NormalizeSockaddrUnPath(sockaddr.Path[:]))
```
It is probably worth noting that this only supports named unix sockets, and we'd need a bit more code to handle abstract sockets, and to separate them from unnamed sockets. That can be a follow-on patch though.

kevsecurity replied:
My bad; didn't read on!

kevsecurity commented on pkg/sensors/tracing/kprobe_net_test.go:
```diff
@@ -1326,3 +1376,148 @@ spec:
+    - index: 1
+      type: "sockaddr_un"
+    selectors:
+    - matchArgs:
+      - index: 0
+        operator: "Family"
+        values: ["AF_UNIX"]
+      - index: 1
+        operator: "Family"
+        values: ["AF_UNIX"]
+      - index: 1
+        operator: "Prefix"
+        values: ["@"]
+    - matchArgs:
+      - index: 0
+        operator: "Family"
+        values: ["AF_UNIX"]
+      - index: 1
+        operator: "Family"
+        values: ["AF_UNIX"]
+      - index: 1
+        operator: "Prefix"
+        values: ["/"]
```
The selectors matching index 1 will match everything because there is no specific matching in bpf/process/types/basic.h.

kevsecurity reviewed (changes requested):
It looks good - thanks! I agree that it would be better divided into a few commits. Did you want to add a case to the switch in selector_arg_offset() (in basic.h) to filter the sockaddr_un parameters (the selectors in your test policy implies you might do)? Let me know if you want help or a chat about that. But otherwise, other than dividing it up, it is looking good.

AshishNaware commented:
I’ve addressed the review comments, and the filters are now working as expected in my local environment. With the current setup, we can filter socket paths by prefix and family. I think this could be extended in the future to support additional path filters (e.g., Equal, NotEqual, Postfix), but I’d prefer to keep that work for a separate PR.

At the moment, I’m working through some pipeline failures and have reached out to @kevsecurity on Slack for guidance.

mtardy reviewed (commented):
> I’ve addressed the review comments, and the filters are now working as expected in my local environment. With the current setup, we can filter socket paths by prefix and family. I think this could be extended in the future to support additional path filters (e.g., Equal, NotEqual, Postfix), but I’d prefer to keep that work for a separate PR.
>
> At the moment, I’m working through some pipeline failures and have reached out to @kevsecurity on Slack for guidance.

Kev will be soon available for review (I would expect next week).

In the meantime, you should be able to now rebase your PR, we fixed something on the generated TracingPolicy docs.

Also if you rebase, if you can be a little bit more specific in your patches titles that would help for when we want to merge this "added tests" or "pr checks" is not sufficient as title/description. Maybe you want to squash the "pr checks" one into other commits above btw.

AshishNaware commented on pkg/selectors/kernel.go:
```diff
@@ -856,8 +856,8 @@ func ParseMatchArg(k *KernelSelectorState, arg *v1alpha1.ArgSelector, sig []v1al
 	case SelectorOpSport, SelectorOpDport, SelectorOpNotSport, SelectorOpNotDport, SelectorOpProtocol, SelectorOpFamily, SelectorOpState:
-		if ty != gt.GenericSockType && ty != gt.GenericSkbType && ty != gt.GenericSockaddrType && ty != gt.GenericSocketType {
-			return errors.New("sock/socket/skb/sockaddr operators specified for non-sock/socket/skb/sockaddr type")
+		if ty != gt.GenericSockType && ty != gt.GenericSkbType && ty != gt.GenericSockaddrType && ty != gt.GenericSocketType && ty != gt.GenericSockaddrUnType {
+			return errors.New("sock/socket/skb/sockaddr/sockaddr_un operators specified for non-sock/socket/skb/sockaddr/sockaddr_un type")
```
resolving this comment as i have added changes to filter out values using selectors.

kevsecurity replied:
Okay, I think this needs a nested if within the if to return an error if sockaddr_un is being used with OpSport, OpDport, etc. Really, it should only be string matchers (if we drop family) so maybe this shouldn't be in this if at all?

AshishNaware commented:
@mtardy / @kevsecurity  - The pipeline is green now. I’ve gated the feature behind the `__LARGE_BPF_PROG` macro to prevent adding excessive instructions on the 4.19 kernel. The change is ready for review.

kevsecurity commented on pkg/api/tracingapi/client_kprobe.go:
```diff
@@ -262,6 +262,11 @@ type MsgGenericKprobeSockaddr struct {
 }
 
+type MsgGenericKprobeSockaddrUn struct {
+	Family uint16
```
For sockaddr_un, Family is always AF_UNIX (according to the man page). Do we need to record it? In inet we do because it tells us whether we're IPv4 or IPv6 (or potentially something we don't understand).

kevsecurity commented on pkg/sensors/tracing/generictracepoint.go:
```diff
@@ -1083,3 +1097,35 @@ func handleMsgGenericTracepoint(
+// NormalizeSockaddrUnPath returns a display-friendly byte slice:
+// "@name" for abstract UNIX sockets (trimming trailing NUL padding),
+// or the pathname up to the first NUL for filesystem sockets.
+func NormalizeSockaddrUnPath(raw []byte) []byte {
```
If we had a flag for abstract sockets then we could use that to spot an abstract socket.

kevsecurity reviewed (changes requested):
This is looking really good. I've left a handful of comments to look at. Happy to discuss them. A bit of a tweak and this can be merged.

AshishNaware commented:
Addressed review comments with below changes-

- Renamed variables and droped redundant checks.
- Modified sockaddr_un_type struct with path length and abstract-socket flag.
- Added support to filter path equal, not equal, prefix, not prefix and family along with tests.

**Pipeline failures:**
e2e tests are successful locally. is there any known pipeline issue? can we retrigger it?

```
go list ./tests/e2e/tests/... | xargs -Ipkg go test  -gcflags= -timeout 20m -failfast -cover pkg  -fail-fast -tetragon.helm.set tetragon.image.override=""cilium/tetragon:latest"" -tetragon.helm.set tetragonOperator.image.override=""cilium/tetragon-operator:latest"" -tetragon.helm.url="" -tetragon.helm.chart="/home/ashish/github.com/cilium/tetragon/install/kubernetes/tetragon" ""
?   	github.com/cilium/tetragon/tests/e2e/tests	[no test files]
ok  	github.com/cilium/tetragon/tests/e2e/tests/labels	289.880s	coverage: [no statements]
ok  	github.com/cilium/tetragon/tests/e2e/tests/policyfilter	129.138s	coverage: [no statements]
ok  	github.com/cilium/tetragon/tests/e2e/tests/skeleton	54.043s	coverage: [no statements]
```

Tests failing on kernels 4.1 seems unrelated to my changes. Any guidance on this?

kevsecurity commented:
> **Pipeline failures:** e2e tests are successful locally. is there any known pipeline issue? can we retrigger it?

I think you might need a rebase to pick up some CI changes.

> Tests failing on kernels 4.1 seems unrelated to my changes. Any guidance on this?

Without looking properly, my guess is that the generic kprobe program no longer loads on 4.19, possibly because it has more than 4096 instructions maybe? Does it work locally?

FedeDP reviewed (commented):
LGTM, left a suggestion and a comment, PTAL 🙏

AshishNaware committed:
api: add unix socket sockaddr_un fields and regenerate CRDs/docs

Changelog:
Extends the public API to surface Unix socket address information in kprobe/tracepoint events.
  - Adds KprobeSockaddrUn protobuf message (family, path) and a sockaddrun_arg oneof variant in KprobeArgument; regenerates all derived Go bindings and event-checker code
  - Adds sockaddr_un as a valid argument type in TracingPolicy CRDs
  - Documents the new message fields and example selectors (Equal, Prefix, NotEqual, NotPrefix, Family) in grpc-api.md and tracing-policy.md

Signed-off-by: Ashish Naware <ashishnaware3@gmail.com>

AshishNaware committed:
bpf: support sockaddr_un parsing and filtering for unix socket events

Changelog:

Implements the full BPF-to-userspace pipeline for sockaddr_un arguments (gated at kernel ≥ 5.11).

  - BPF: new sockaddr_un.h header with struct sockaddr_un_type and set_event_from_sockaddr_un() which reads the kernel struct, detects abstract sockets (leading \0), strips it, and stores the printable path and length. Adds copy_sockaddr_un() for ring buffer copy and filter_sockaddr_un() supporting Prefix, NotPrefix, Equal, NotEqual, and Family operators. All wired into the existing type/filter dispatch tables.

  - Userspace: registers the new type, adds decodeSockaddrUnPath() (prepends @ for abstract names), decodes it in both kprobe and tracepoint paths, and converts to protobuf in the gRPC layer. Selector validation blocks port/protocol/state operators on sockaddr_un and restricts them to string and Family operators only.

Signed-off-by: Ashish Naware <ashishnaware3@gmail.com>

AshishNaware committed:
tests: added unit tests for tracing unix socket events

Changelog:
Adds an integration test suite (KprobeNet) in kprobe_net_test.go with helpers startUnixServer and dialUnix, covering five selector scenarios against both filesystem and abstract Unix sockets: Prefix, Equal, NotEqual, NotPrefix, and Family-only. All tests are gated at kernel ≥ 5.11.

Signed-off-by: Ashish Naware <ashishnaware3@gmail.com>

AshishNaware commented on bpf/process/types/sockaddr_un.h:
```diff
@@ -0,0 +1,58 @@
+	memset(event, 0, sizeof(*event));
+	if (probe_read(&event->family, sizeof(event->family), _(&address->sa_family)) < 0)
+		return;
+	probe_read(event->sun_path, sizeof(event->sun_path), un->sun_path);
```
@FedeDP - that was a good catch, thanks for the review. I have handled the probe_read error in recent changes. Also implemented your suggestion regarding sending real bytes to userspace.

FedeDP reviewed (approved):
LGTM! Thank you very much :) Let's wait for @kevsecurity 🚀

kevsecurity reviewed (approved):
🎉  Looks good!

olsajiri commented on pkg/sensors/tracing/kprobe_net_test.go:
```diff
@@ -1326,3 +1377,311 @@ spec:
+	const absSock = "@tetragon-abs"
+
+	cleanupFS := startUnixServer(suite.T(), fsSock)
+	defer cleanupFS()
```
curious, could this be just `defer startUnixServer(suite.T(), fsSock)` ?

olsajiri reviewed (commented):
looks good, left some comments, please provide decriptive changelogs to each commit, thanks

mtardy reviewed (commented):
can you update the docs with that new type in a sep commit? https://tetragon.io/docs/concepts/tracing-policy/argument_types/

AshishNaware committed:
docs: add sockaddr_un argument type documentation

docs: document `sockaddr_un` argument type for Unix domain socket kprobe filtering

Signed-off-by: Ashish Naware <ashishnaware3@gmail.com>

mtardy reviewed (approved):
docs look ok! thanks

olsajiri commented on bpf/process/types/sockaddr_un.h:
```diff
@@ -0,0 +1,54 @@
+		max_len = UNIX_SOCKET_PATH_MAX - 1;
+	}
+	if (probe_read(event->sun_path, max_len, src_path) < 0)
+		return;
```
hum, with leading NULL byte the length is zero also with your current code, could you please explain in more detail what's the issue? thnx

AshishNaware replied:
Sorry, I was wrong to reason that `probe_read_str` may not work due to leading `NUL` byte as the `probe_read` is reading full path (or from `sun_path + 1` in case of abstract sockets) until the first `NUL` byte is detected.

As @kevsecurity pointed out in current thread and [past comment ](https://github.com/cilium/tetragon/pull/3980#discussion_r2251003784) `probe_read_str()` may not work in case of abstract sockets not ending with `NUL`. 

However, after reading below comment from the codebase, 
```
// probe_read_str() always nul-terminates the string.
// So add one to the length to allow for it. This should
// result in us honouring our max_size correctly.
```
I am kind of thinking whether probe_read_str() is a good choice as when there exists an abstract path (without `NUL` terminator)  i.e. `NUL + 107 bytes`, it would just return a path of size 108. (`107 bytes + NUL`)

kevsecurity replied:
When the man page (https://man7.org/linux/man-pages/man7/unix.7.html) states:

```
Null bytes in the name have no special significance.
```

does this mean the abstract path can contain null bytes? I read it that it does, so a probe_read_str() that stops on a null byte would be wrong IMO. Also, if it does not, then why not specify the length using strnlen() (similar to what the normal socket path description does, but with a max length, n)? I think this is strong evidence that the abstract path can contain null characters.

AshishNaware replied:
> The name of the socket is contained in the first (addrlen - sizeof(sa_family_t)) bytes of sun_path. 

So we need addrlen to extract the actual path and use `probe_read` instead of `probe_read_str` (since null bytes can appear in the name). Any suggestions on how to obtain addrlen?

kevsecurity replied:
Often when a function takes a sockaddr, it also takes a len. We have other data types (one I think) in Tetragon where they require a length parameter (memory pointer, for example). I'll take a look at the code next week and see if we can do the same thing.

mtardy replied:
So I'm not exactly sure on how this ended up being fixed, related comment here as well So I'm not exactly sure on how this ended up being fixed.

Could you clarify what what was decided?

kevsecurity replied:
So, unfortunately I think ```probe_read_str()``` is wrong because the data could contain nuls. The best solution would be to ```probe_read_kernel()``` with a length that is 3 less than the supplied length param (2 less because that is the size of the family parameter, plus 1 for the leading nul), but that adds the complexity of needing the length parameter.

I think maybe the easiest answer here is to ```probe_read_kernel()``` the full 107 byte buffer (starting from offset 1) and deal with it in user space. I imagine we might not display abstract paths with nuls correctly at this point anyway. Maybe if we make this small change (probe_read_str -> probe_read_kernel with len 107), then we can merge and create an issue for someone to look at abstract sockets more closely?

olsajiri replied:
> I think maybe the easiest answer here is to probe_read_kernel() the full 107 byte buffer (starting from offset 1) and deal with it in user space. I imagine we might not display abstract paths with nuls correctly at this point anyway. Maybe if we make this small change (probe_read_str -> probe_read_kernel with len 107), then we can merge and create an issue for someone to look at abstract sockets more closely?

sounds good

olsajiri commented on pkg/grpc/tracing/tracing.go:
```diff
@@ -634,6 +644,16 @@ func (msg *MsgGenericTracepointUnix) HandleMessage() *tetragon.GetEventsResponse
+		case tracingapi.MsgGenericKprobeArgSockaddrUn:
+			address := tetragon.KprobeSockaddrUn{
+				Family: familyString(v.Family),
+				Path:   strutils.UTF8FromBPFBytes([]byte(v.Path)),
```
is this path already converted?

AshishNaware replied:
yep. fixed it. 

I see the CI is failing on windows build. Can we re-trigger the workflows if those are flaky tests? I don't think it has something to do with this fix.

olsajiri reviewed (commented):
left one comment, otherwise lgtm, thanks

FedeDP commented:
I am not sure whether the test:
```
--- FAIL: Test_rateLimitExport (0.05s)
    --- FAIL: Test_rateLimitExport/rate_limit_(100_events,_10_rate_limit) (0.05s)
        exporter_test.go:277: 
            	Error Trace:	D:/a/tetragon/tetragon/go/src/github.com/cilium/tetragon/pkg/exporter/exporter_test.go:196
            	            				D:/a/tetragon/tetragon/go/src/github.com/cilium/tetragon/pkg/exporter/exporter_test.go:277
            	Error:      	Not equal: 
            	            	expected: 10
            	            	actual  : 12
            	Test:       	Test_rateLimitExport/rate_limit_(100_events,_10_rate_limit)
            	Messages:   	number of events
        exporter_test.go:277: 
            	Error Trace:	D:/a/tetragon/tetragon/go/src/github.com/cilium/tetragon/pkg/exporter/exporter_test.go:198
            	            				D:/a/tetragon/tetragon/go/src/github.com/cilium/tetragon/pkg/exporter/exporter_test.go:277
            	Error:      	Not equal: 
            	            	expected: 0x5a
            	            	actual  : 0x58
            	Test:       	Test_rateLimitExport/rate_limit_(100_events,_10_rate_limit)
            	Messages:   	number of dropped events
FAIL
FAIL	github.com/cilium/tetragon/pkg/exporter	0.380s
```
is flaky; never seen that failure (@mtardy do you?).

mtardy commented:
> is flaky; never seen that failure (@mtardy do you?).

never seen that one in particular nop, looks like it was flaky on windows also

mtardy commented on bpf/process/types/sockaddr_un.h:
```diff
@@ -0,0 +1,42 @@
+#ifndef __SOCKADDR_UN_H__
+#define __SOCKADDR_UN_H__
+
+#define UNIX_SOCKET_PATH_MAX 108
+
+struct sockaddr_un_type {
+	__u16 family;
+	bool is_abstract;
+	__u8 path_len;
+	char sun_path[UNIX_SOCKET_PATH_MAX];
+};
+
+/* set_event_from_sockaddr_un(event, address)
+ *
+ * Populate the event args with the sock info.
+ */
+FUNC_INLINE void
+set_event_from_sockaddr_un(struct sockaddr_un_type *event, struct sockaddr *address)
+{
+	struct sockaddr_un *un = (struct sockaddr_un *)address;
+	char first = 0;
+	const char *src_path = un->sun_path;
+	int ret;
+
+	memset(event, 0, sizeof(*event));
+	if (probe_read(&event->family, sizeof(event->family), _(&address->sa_family)) < 0)
+		return;
+	if (probe_read(&first, sizeof(first), src_path) < 0)
+		return;
+	event->is_abstract = (first == '\0');
+	if (event->is_abstract)
+		src_path = un->sun_path + 1;
+
+	ret = probe_read_str(event->sun_path, UNIX_SOCKET_PATH_MAX, src_path);
+	if (ret < 0)
+		return;
+	event->path_len = (__u8)(ret - 1);
+}
```
So just before merging, could we clarify that this satisfies everyone, I was not sure of the outcome of this discussion https://github.com/cilium/tetragon/pull/3980#discussion_r3008914504.

cc @kevsecurity @olsajiri

mtardy replied:
From offline discussions I think we can always open a separate issue to take a closer look on this if needed but we can move forward with this PR.

mtardy replied:
Also for next time Ashish, you don't need to add "changelog" to your git commit description, and please wrap the descriptions so that it fits ~80 chars.

kevsecurity replied:
This looks great! Thanks.

kevsecurity commented:
I think you might only need a rebase to get through the CI. Once tests pass, I'm happy to merge.
